### PR TITLE
Feature/3098 enhance nonce messaging

### PR DESCRIPTION
**Pull request by ivankruchkoff**
## Summary

Addresses issue #3098 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Core/Authentication/Authentication.php [hunk 1]
```diff
@@ -1353,4 +1353,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
 	}
 
+	/**
+	 * Invalid nonce error handler.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $action    Action name.
+	 */
+	public function invalid_nonce_error( $action ) {
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1363]: Let's make this method to be static. In this case, we won't need to instantiate a new instance of the Authentication class in the Reset class.

File: includes/Core/Authentication/Authentication.php [hunk 2]
```diff
@@ -1353,4 +1353,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
+		if ( strpos( $action, 'googlesitekit_proxy_' ) === 0 ) {
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1364]: Let's return early if the action doesn't start with `googlesitekit_proxy_`:

```php
if ( strpos( $action, 'googlesitekit_proxy_' ) === false ) {
    wp_nonce_ays( $action );
    return;
}

// Copied from wp_nonce_ays() with tweak to the url.
$html = __( 'The link you followed has expired.', 'default' );
...
```

File: includes/Core/Authentication/Authentication.php [hunk 3]
```diff
@@ -1353,4 +1353,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
 		return $feature_enabled;
 	}
 
+	/**
+	 * Invalid nonce error handler.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $action    Action name.
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1361]:
```suggestion
	 * @param string $action Action name.
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @ivankruchkoff. This mostly looks good to me. Added a few minor comments. Once they are addressed, we should be good to go.

File: includes/Core/Authentication/Authentication.php (hunk 2 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/ef254b84ba25533d0a89d04a5b637a3450a5ebe5

File: includes/Core/Authentication/Authentication.php (hunk 1 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php:1369]: If we make it static, we will need to pass $context param to call `$this->context->admin_url( 'splash' )` later in the function.

Comment (eugene-manuilov) [reply includes/Core/Authentication/Authentication.php:1369]: We can use `Plugin::instance()->context()->admin_url( ... )`. It's better than instantiating a new instance of the `Authentication` class in the `Reset.php` file.

Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/bd06a5b10d75b7060778cd1ce3d2f38e076790c8

File: includes/Core/Authentication/Authentication.php [hunk 4]
```diff
@@ -1359,4 +1360,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $action Action name.
+	 */
+	public static function invalid_nonce_error( $action ) {
+		if ( strpos( $action, 'googlesitekit_proxy_' ) !== 0 ) {
+			wp_nonce_ays( $action );
+			return;
+		} else {
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1374]: This is redundant, we don't need this `else` here. Let's get rid of it.

File: includes/Core/Authentication/Authentication.php [hunk 5]
```diff
@@ -1359,4 +1360,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
+			// Copied from wp_nonce_ays() with tweak to the url.
+			$html  = __( 'The link you followed has expired.', 'default' );
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1376]: Even though the translation is available in the default text domain, we need to use our own text domain here to make sure we have translations even when this string is changed in the core.

```suggestion
			$html  = __( 'The link you followed has expired.', 'google-site-kit' );
```

File: includes/Core/Authentication/Authentication.php [hunk 6]
```diff
@@ -1359,4 +1360,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
+			$html .= '</p><p>';
+			$html .= sprintf(
+				'<a href="%s">%s</a>',
+				esc_url( Plugin::instance()->context()->admin_url( 'splash' ) ),
+				__( 'Please try again.', 'default' )
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1381]:
```suggestion
				__( 'Please try again.', 'google-site-kit' )
```

File: includes/Core/Authentication/Authentication.php [hunk 7]
```diff
@@ -1359,4 +1360,27 @@ private function filter_features_via_proxy( $feature_enabled, $feature_name ) {
...
+			);
+			wp_die( $html, __( 'Something went wrong.', 'default' ), 403 ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1383]:
```suggestion
			wp_die( $html, __( 'Something went wrong.', 'google-site-kit' ), 403 ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```

File: tests/phpunit/integration/Core/Authentication/AuthenticationTest.php [hunk 8]
```diff
@@ -888,6 +888,25 @@ function ( $preempt, $args, $url ) use ( $google_proxy ) {
...
+		} catch ( WPDieException $exception ) {
+			$this->assertStringStartsWith( 'You are attempting to log out of Test Blog', $exception->getMessage() );
+			return;
+		}
+		try {
+			Authentication::invalid_nonce_error( 'googlesitekit_proxy_foo_action' );
+		} catch ( WPDieException $exception ) {
+			$this->assertEquals( 'The link you followed has expired.</p><p><a href="http://example.org/wp-admin/admin.php?page=googlesitekit-splash">Please try again.</a>', $exception->getMessage() );
+			return;
+		}
```
Comment (eugene-manuilov) [inline tests/phpunit/integration/Core/Authentication/AuthenticationTest.php:903]: Let's split it into two tests.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @ivankruchkoff. Added a few more final comments. Once they are addressed, we should be good to go.

File: includes/Core/Authentication/Authentication.php (hunk 4 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/72855cff21aa68eb4977874409f63f5ca88292a4

File: includes/Core/Authentication/Authentication.php (hunk 5 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/72855cff21aa68eb4977874409f63f5ca88292a4

File: includes/Core/Authentication/Authentication.php (hunk 6 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/72855cff21aa68eb4977874409f63f5ca88292a4

File: includes/Core/Authentication/Authentication.php (hunk 7 above)
Comment (ivankruchkoff) [reply includes/Core/Authentication/Authentication.php]: https://github.com/google/site-kit-wp/pull/3976/commits/72855cff21aa68eb4977874409f63f5ca88292a4

File: tests/phpunit/integration/Core/Authentication/AuthenticationTest.php (hunk 8 above)
Comment (ivankruchkoff) [reply tests/phpunit/integration/Core/Authentication/AuthenticationTest.php]: https://github.com/google/site-kit-wp/pull/3976/commits/8fe8aba732485d21e1219f969c254c81d018d0da

**Review by eugene-manuilov — APPROVED** [review]
LGTM

**Pull request merged**